Scripts/Runtime/Collections/DeferredNativeArray.cs

```diff
@@ -0,0 +1,333 @@
+        }
+
+        private static unsafe void Allocate(Allocator allocator, out DeferredNativeArray<T> array)
+        {
+            if (allocator <= Allocator.None)
+            {
+                throw new ArgumentException("Allocator must be Temp, TempJob or Persistent", nameof(allocator));
+            }
+
+            IsUnmanagedAndThrow();
+            array = new DeferredNativeArray<T>();
+            void* initialBuffer = UnsafeUtility.Malloc(SIZE, ALIGNMENT, allocator);
+            array.m_BufferInfo = (DeferredNativeArrayBufferInfo*)UnsafeUtility.Malloc(DeferredNativeArrayBufferInfo.SIZE,
+                                                                                      DeferredNativeArrayBufferInfo.ALIGNMENT,
+                                                                                      allocator);
+            array.m_BufferInfo->Length = 0;
+            array.m_BufferInfo->MaxIndex = 0;
+            array.m_BufferInfo->Buffer = initialBuffer;
+            array.m_BufferInfo->State = DeferredNativeArrayState.Placeholder;
+
+            array.m_Allocator = allocator;
+
+            DisposeSentinel.Create(out array.m_Safety, out array.m_DisposeSentinel, 1, allocator);
+            InitStaticSafetyId(ref array.m_Safety);
+        }
+
+        // ReSharper disable once StaticMemberInGenericType
+        private static int s_StaticSafetyId;
+
+        [NativeDisableUnsafePtrRestriction] internal unsafe DeferredNativeArrayBufferInfo* m_BufferInfo;
+
+        [NativeSetClassTypeToNullOnSchedule] private DisposeSentinel m_DisposeSentinel;
+
+        internal AtomicSafetyHandle m_Safety;
+        private Allocator m_Allocator;
+
+        /// <summary>
+        /// Whether the collection has been created or not.
+        /// </summary>
+        public unsafe bool IsCreated
+        {
+            get => m_BufferInfo != null && m_BufferInfo->Buffer != null;
+        }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="DeferredNativeArray{T}"/>
+        /// </summary>
+        /// <param name="allocator">The <see cref="Allocator"/> to use for memory allocation.</param>
+        public DeferredNativeArray(Allocator allocator)
+        {
+            Allocate(allocator, out this);
+        }
+
+        /// <summary>
+        /// Disposes the collection
+        /// </summary>
+        [WriteAccessRequired]
+        public unsafe void Dispose()
+        {
+            if (!UnsafeUtility.IsValidAllocator(m_Allocator))
+            {
+                throw new InvalidOperationException("The DeferredNativeArray can not be Disposed because it was not allocated with a valid allocator.");
+            }
+
+            DisposeSentinel.Dispose(ref m_Safety, ref m_DisposeSentinel);
+
+            if (m_BufferInfo == null)
+            {
+                return;
+            }
+
+            UnsafeUtility.Free(m_BufferInfo->Buffer, m_Allocator);
```

Comment:
Hmmn, I suppose we could. 

Unity doesn't in any of their collections. They seem to treat all their internal structs and just Plain Old Data structs and they allocate and free from outside. 

Thoughts?

Reply:
We discussed and we're going to leave as it. 

There are a couple of reasons that make it more tricky to have the allocate and free inside. 

1. We can't actually call a constructor on the struct because we are creating the struct via `void* initialBuffer = UnsafeUtility.Malloc(SIZE, ALIGNMENT, allocator);`
2. If we moved the `UnsafeUtility.Free` inside a Dispose method, we'd have to pass in the allocator as well which exists outside. 

None of these are showstoppers, we could do it but it would be a little bit odd and it's probably easier to keep it like this to reference for similarities with Unity's collections. 

Although, to be fair, Unity's collections are littered with inconsistencies themselves.